### PR TITLE
Token-aware condenser

### PR DESCRIPTION
**Pull request by enyst**
- [ ] This change is worth documenting at https://docs.all-hands.dev/
- [ ] Include this change in the Release Notes. If checked, you **must** provide an **end-user friendly** description for your change below

---
**Give a summary of what the PR does, explaining any non-trivial design decisions.**

This PR proposes to add a token-aware LLM summarizing condenser.

---
**Link of any specific issues this addresses.**

### CODE REVIEW

**Comment by enyst** [comment]
@openhands-agent Use the Github API to read the diff of this PR, PR 7587. Understand what it aims to achieve.

Add unit tests in test_condenser.py, make sure to use as example the existing tests, so that we have good coverage.

Run test_condenser.py to make sure they pass.

**Comment by openhands-agent** [comment]
[OpenHands](https://github.com/All-Hands-AI/OpenHands) started fixing the pr! You can monitor the progress [here](https://github.com/All-Hands-AI/OpenHands/actions/runs/14140416701).

**Comment by openhands-agent** [comment]
[OpenHands](https://github.com/All-Hands-AI/OpenHands) started fixing the pr! You can monitor the progress [here](https://github.com/All-Hands-AI/OpenHands/actions/runs/14145317248).

**Comment by openhands-agent** [comment]
Here's a concise overview of the changes and their impact:

RESOLVED:
✓ Replaced artificial mocking with proper token limit testing
✓ Added realistic condensation triggers
✓ Improved verification coverage
✓ Enhanced test observability
✓ Fixed RollingCondenserTestHarness functionality

The changes appear to have successfully transformed the test into a more authentic and comprehensive validation of the TokenAwareCondenser's behavior. The test now properly verifies:
- Natural view growth
- Condensation timing
- Observation handling
- Size management
- Multiple condensation cycles

ASSESSMENT:
Based on the information provided, all major issues appear to have been successfully resolved. The test is now both more realistic and more thorough in validating the condenser's actual behavior.

No remaining issues are apparent from the provided information, though it might be worth confirming that edge cases and error conditions are also covered in the test suite.

File: openhands/memory/condenser/impl/token_aware_condenser.py [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+    Condensation,
+    RollingCondenser,
+    View,
+)
+
+
+class TokenAwareCondenser(RollingCondenser):
+    """A condenser that summarizes events when they exceed a token limit.
+
+    Maintains a condensed history and summarizes events when they exceed a token threshold,
```
Comment (enyst) [inline openhands/memory/condenser/impl/token_aware_condenser.py:20]:
```suggestion
    Similar with LLMSummarizingCondenser, with token limits.
    Maintains a condensed history and summarizes events when they exceed a token threshold,
```

**Comment by enyst** [comment]
@openhands-agent Read carefully the diff of this PR, PR 7587. You can use the GitHub API with the github token you have in env to read it. Understand what it aims to achieve.

TASK:
The unit test test_condenser.py::test_token_aware_condenser_integration is failing. Please fix it. 

The error is:
> Verify the completion method was called for each condensation
>           assert mock_llm.completion.call_count == condensation_count
> E           AssertionError: assert 22 == 0
> E            +  where 22 = <MagicMock name='mock.completion' id='[139685889619200](tel:139685889619200)'>.call_count
> E            +    where <MagicMock name='mock.completion' id='[139685889619200](tel:139685889619200)'> = <MagicMock spec='LLM' id='[139685889607824](tel:139685889607824)'>.completion
> tests/unit/test_condenser.py:790: AssertionError

I suggest to try to reset mock_llm or store its count at the beginning of the test, otherwise it counts all calls across all tests in the file (that's how it got 22!).

Make sure the test continues to actually test what we need to test, don't cheat!

You only need to fix this test, and run the test_condenser.py test file. You don't need to run others.

**Comment by openhands-agent** [comment]
[OpenHands](https://github.com/All-Hands-AI/OpenHands) started fixing the pr! You can monitor the progress [here](https://github.com/All-Hands-AI/OpenHands/actions/runs/14145915602).

**Comment by openhands-agent** [comment]
Here's a concise overview:

Status: Issues Not Resolved

Changes Made:
- Added mock reset functionality
- Implemented new event counting mechanism
- Improved edge case handling for empty events

Remaining Issues:
- Mock call count (22) still doesn't match expected count (0)
- Mock reset may be ineffective
- Event counting remains unsynchronized with LLM calls

Summary: While the changes addressed some structural issues, the core problem of mismatched mock calls persists. Further investigation is needed to properly synchronize the mock behavior with the expected test outcomes.

**Comment by enyst** [comment]
@openhands-agent-exp

This branch has merge conflicts with main. Please fix them. Your environment is already set up on this PR.

You should run pre-commit when you're done. Do not run any tests, I took care of those in CI.

**Comment by openhands-agent** [comment]
[OpenHands](https://github.com/All-Hands-AI/OpenHands) started fixing the pr! You can monitor the progress [here](https://github.com/All-Hands-AI/OpenHands/actions/runs/14815440743).

**Comment by enyst** [comment]
@openhands-agent-exp Fix the merge conflicts of this PR branch with main.

You do not need to run tests, I already took care of those in CI. Just be careful with merges and give me a full report in your Finish action when you're done.

**Comment by openhands-agent** [comment]
[OpenHands](https://github.com/All-Hands-AI/OpenHands) started fixing the pr! You can monitor the progress [here](https://github.com/All-Hands-AI/OpenHands/actions/runs/15354709907).

**Comment by Truenya** [comment]
@enyst great job! Big thanks, will it be merged soon?

**Comment by enyst** [comment]
Hi, I'm OpenHands-GPT-4.1, assisting on this PR.

I’ve pushed updates to branch enyst/used-tokens after completing merge conflict resolution and running tests locally.

PR details (via GitHub API):
- PR #: 7587
- Title: Token-aware condenser
- State: open (not draft)
- Base: All-Hands-AI:main (ref: main, sha: e40681c…)
- Head: All-Hands-AI:enyst/used-tokens (sha: 62b43e4f3…)
- Changed files: 9
- Commits: 30
- Additions/Deletions: +500 / -27
- URL: https://github.com/All-Hands-AI/OpenHands/pull/7587

What I did locally:
- Completed the in-progress rebase and resolved remaining conflicts.
- Fixed signatures and removed conflict markers in: observation_masking_condenser.py and recent_events_condenser.py.
- Restored message_utils.get_token_usage_for_event to prefer event.response_id with fallback to tool_call_metadata.model_response.id.
- Standardized tests to patch exceeds_token_limit in TokenAwareCondenser.
- Adjusted RollingCondenserTestHarness to append condensation events to state.history.
- Cleaned tests/unit/test_condenser.py (removed duplicate/incomplete test, stray markers).
- Verified tests/unit/test_condenser.py: 35 passed locally.

Let me know if you'd like me to update the PR description with a summary of TokenAware behavior and the message_utils decision order.
